### PR TITLE
fixing custom dates bug

### PR DESCRIPTION
**Pull request by alagroy-42**
#1810 but on release3_3

### CODE REVIEW

**Comment by nboisteault** [comment]
@alagroy-42 I have tested and saving custom date is ok. Now I'd like to see the date being filled in the date picker when editing the feature. Thanks.

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 1]
```diff
@@ -459,6 +459,25 @@ public function setFormDataFromFields($feature)
...
+     *
+     * @param string $value The datetime to convert
+     * @param object $edittype The format of the date
+     */
+    public function convertDateTimeToFormat($value, $edittype)
+    {
+        if (!property_exists($edittype, 'options') || !property_exists($edittype->options, 'field_format')) {
+            return $value;
+        }
+        $dateTab = explode('-', $value);
```
Comment (nboisteault) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:473]: You have the code handling this specific date format (yyyyMMdd) but I'd like it to be more generic and handle other formats like in this screenshot showing QGIS customizing date part.

image

**Review by nboisteault — CHANGES_REQUESTED** [review]

**Comment by Gustry** [comment]
@alagroy-42 Side note, can you also rename the commit with the actual bugfix ? You can skip the name of the customer I think. Thanks

**Comment by josemvm** [comment]
https://github.com/3liz/lizmap-web-client/issues/1076

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 2]
```diff
@@ -459,6 +459,25 @@ public function setFormDataFromFields($feature)
...
+        $value = $edittype->options->field_format;
+        $value = str_replace('yyyy', $dateTab[0], $value);
+        $value = str_replace('MM', $dateTab[1], $value);
+
+        return str_replace('dd', $dateTab[2], $value);
```
Comment (rldhont) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:478]:
```suggestion
        $date = new DateTime($value);
        return $date->format($edittype->options->field_format);
```
I think this is the easiest way to do. https://www.php.net/manual/fr/datetime.construct.php

**Review by rldhont — CHANGES_REQUESTED** [review]
Is it possible to add a test for the method `convertDateTimeToFormat` ?

**Review by rldhont — APPROVED** [review]

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 3]
```diff
@@ -470,12 +470,44 @@ public function convertDateTimeToFormat($value, $edittype)
         if (!property_exists($edittype, 'options') || !property_exists($edittype->options, 'field_format')) {
             return $value;
         }
-        $dateTab = explode('-', $value);
-        $value = $edittype->options->field_format;
-        $value = str_replace('yyyy', $dateTab[0], $value);
-        $value = str_replace('MM', $dateTab[1], $value);
+        $dateFormat = $edittype->options->field_format;
+        // convertion from QGIS to PHP format
```
Comment (nboisteault) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:474]: convertion => conversion

**Comment by nboisteault** [comment]
@alagroy-42 
Tested **OK** : 

- yyyyMMdd
- ddMMyyyy
- yyyy.MM.dd
- ddd MMMM d yy
- yyyy-MM-dd HH:mm

Tested **NOK** : 
- h:m:s ap
In this case, we only get the datepicker without timepicker. '12:00:00 am' is saved though.

**Comment by rldhont** [comment]
> Tested **NOK** :
> 
>     * `h:m:s ap`
>       In this case, we only get the datepicker without timepicker. '12:00:00 am' is saved though.
@nboisteault time control is not supported in Lizmap-web-client 3.3, so I think that the fix is ended.

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 4]
```diff
@@ -452,13 +452,70 @@ public function setFormDataFromFields($feature)
                 }
                 $form->setData($ref.'_hidden', $value);
             } else {
+                if (in_array(strtolower($this->formControls[$ref]->fieldEditType), array('date', 'time', 'datetime'))) {
+                    $date = new DateTime($value);
```
Comment (rldhont) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:456]: $value could be in the `$edittype->options->field_format`

**Review by rldhont — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/classes/qgisForm.class.php (hunk 4 above)
Comment (alagroy-42) [reply lizmap/modules/lizmap/classes/qgisForm.class.php:456]: Here, `$value` is the date/time already stored in the DB that we want to put in the datepicker when we edit the feature and to do so, we need to convert it into the format accepted by Jelix.
[date](https://github.com/jelix/jelix/blob/8f313e8aa19f7a72bad7733a4f9354f5916b9ccb/lib/jelix/plugins/formwidget/date_html/date_html.formwidget.php#L79)
[datetimepicker](https://github.com/jelix/jelix/blob/8f313e8aa19f7a72bad7733a4f9354f5916b9ccb/lib/jelix/plugins/formwidget/datetime_html/datetime_html.formwidget.php#L89)

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 5]
```diff
@@ -452,13 +452,188 @@ public function setFormDataFromFields($feature)
...
+                if (in_array(strtolower($this->formControls[$ref]->fieldEditType), array('date', 'time', 'datetime'))) {
+                    $edittype = $this->formControls[$ref]->getEditType();
+                    if (!$edittype || !property_exists($edittype, 'options')
+                        || !property_exists($edittype->options, 'field_format')) {
+                        break ;
+                    }
+                    $format = $this->convertQgisFormatToPHP($edittype->options->field_format);
+                    $date = DateTime::createFromFormat($format, $value);
+                    $value = $date->format('Y-m-d H:i:s');
+                }
```
Comment (rldhont) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:464]:
```suggestion
                if (in_array(strtolower($this->formControls[$ref]->fieldEditType), array('date', 'time', 'datetime'))) {
                    $edittype = $this->formControls[$ref]->getEditType();
                    if ($edittype && property_exists($edittype, 'options')
                        && property_exists($edittype->options, 'field_format')) {
                        $format = $this->convertQgisFormatToPHP($edittype->options->field_format);
                        $date = DateTime::createFromFormat($format, $value);
                        $value = $date->format('Y-m-d H:i:s');
                    }
                }
```
I was not sure that you can use a `break` here. @laurentj what do you think about it ?

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 6]
```diff
@@ -452,13 +452,188 @@ public function setFormDataFromFields($feature)
...
                 $form->setData($ref, $value);
             }
         }
 
         return $form;
     }
 
+    public function check($feature = null)
```
Comment (rldhont) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:472]: Remove this method that comes from master and is not used in 3.3

**Review by rldhont — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 7]
```diff
@@ -452,13 +452,85 @@ public function setFormDataFromFields($feature)
...
+
+        return $dateFormat;
+    }
+
+    /**
+     * Converts the datetime to the format specified in the qgis Project.
+     *
+     * @param string $value       The datetime to convert
+     * @param object $edittype    The format of the date
+     * @param mixed  $fieldFormat
```
Comment (rldhont) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:523]:
```suggestion
     * @param string $value           The datetime to convert
     * @param string $fieldFormat The format of the date
     * @return string
```
Update the docstring

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 8]
```diff
@@ -452,13 +452,85 @@ public function setFormDataFromFields($feature)
...
+            'AP' => 'A',
+            'ap' => 'a',
+            'm' => 'i',
+            'mm' => 'i',
+            'ss' => 's',
+            't' => 'T',
+        ));
+    }
+
+    public function convertQgisFormatToPHP($fieldFormat)
```
Comment (rldhont) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:497]: Can you add a docstring ?

**Review by rldhont — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/classes/qgisForm.class.php [hunk 9]
```diff
@@ -452,13 +452,85 @@ public function setFormDataFromFields($feature)
                 }
                 $form->setData($ref.'_hidden', $value);
             } else {
+                if (in_array(strtolower($this->formControls[$ref]->fieldEditType), array('date', 'time', 'datetime'))) {
+                    $edittype = $this->formControls[$ref]->getEditType();
+                    if ($edittype && property_exists($edittype, 'options')
+                            && property_exists($edittype->options, 'field_format')) {
+                        $format = $this->convertQgisFormatToPHP($edittype->options->field_format);
+                        $date = DateTime::createFromFormat($format, $value);
+                        $value = $date->format('Y-m-d H:i:s');
```
Comment (nboisteault) [inline lizmap/modules/lizmap/classes/qgisForm.class.php:461]: I have an error `Call to a member function format() on bool` when `$value` is `null`. Please fix.

**Review by nboisteault — CHANGES_REQUESTED** [review]

**Review by rldhont — APPROVED** [review]

**Review by nboisteault — APPROVED** [review]
Thanks!

**Comment by josemvm** [comment]
before:
imagem


after:
imagem -- must be --> 2020-10-02

@alagroy-42 thanks for your great work on LWC

**Pull request merged**